File: neetcode/solutions/no_category/sol347.js
// 347. Top K Frequent Elements
// https://leetcode.com/problems/top-k-frequent-elements/

// broken:
var topKFrequent = function(nums, k) {
    // create a hashmap with num: numOccurrences
    let hashmp = {}
    for (let i=0; i<nums.length; i++){
        if (Object.keys(hashmp).includes(nums[i].toString())){
            hashmp[nums[i]] += 1
        } else {
            hashmp[nums[i]] = 1
        }
    }
    console.log(Object.keys(hashmp))
    console.log(hashmp)
    // sort hashmap by numOccurrences
    let sorted = []
    for (let entry in hashmp){
        sorted.push([entry, hashmp[entry]])
    }
    sorted.sort((a,b)=>{a[1]-b[1]}) // I believe the bug is here, not sorting by value correctly
    console.log(sorted)
    sorted = sorted.slice(0,k)
    let result = []
    sorted.forEach((item)=> {result.push(parseInt(item[0]))})
    return result
    // return an array of k highest
};


// working solution from leetcode solutions
var topKFrequent = function(nums, k) {
    const freqMap = new Map()
    const bucket = []
    const result = []

    for (let num of nums) {
        freqMap.set(num, (freqMap.get(num) || 0)+1)
    }
    // console.log(freqMap)
    for (let [num, freq] of freqMap){
        bucket[freq] = (bucket[freq] || new Set()).add(num)
    }
    // console.log(bucket)
    for (let i=bucket.length-1; i>= 0; i--){
        if(bucket[i]) {result.push(...bucket[i])}
        if (result.length === k) break;
    }
    return result
};

File: neetcode/second_pass/sol_347.py
from typing import List
class Solution():
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashMap = {}
        freq = [[]for i in range(len(nums) + 1)]
        for num in nums:
            hashMap[num] = hashMap.get(num, 0) + 1
        for num, count in hashMap.items():
            freq[count].append(num)
        res = []
        for i in range(len(freq) - 1, 0, -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res




class Solution2():
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashMap = {} # value: count
        for num in nums:
            hashMap[num] = hashMap.get(num, 0) + 1

        buckets = [[]for i in range(len(nums) + 1)]
        for num, count in hashMap.items():
            buckets[count].append(num)

        res = []
        for i in range(len(buckets) - 1, -1, -1):
            for num in buckets[i]:
                res.append(num)
                if len(res) == k:
                    print(res)
                    return res

inst = Solution2()
inst.topKFrequent(nums=[5,5,5,4,4,3], k=2)
